components/feature/downloads/src/main/java/mozilla/components/feature/downloads/DownloadMiddleware.kt

```diff
@@ -7,34 +7,100 @@ package mozilla.components.feature.downloads
 import android.app.DownloadManager
 import android.content.Context
 import android.content.Intent
+import androidx.annotation.VisibleForTesting
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.InternalCoroutinesApi
+import kotlinx.coroutines.flow.collect
+import kotlinx.coroutines.launch
 import mozilla.components.browser.state.action.BrowserAction
 import mozilla.components.browser.state.action.DownloadAction
 import mozilla.components.browser.state.state.BrowserState
+import mozilla.components.browser.state.state.content.DownloadState.Status.CANCELLED
+import mozilla.components.browser.state.state.content.DownloadState.Status.COMPLETED
 import mozilla.components.lib.state.Middleware
 import mozilla.components.lib.state.MiddlewareStore
+import mozilla.components.support.base.log.logger.Logger
+import kotlin.coroutines.CoroutineContext
 
 /**
  * [Middleware] implementation for managing downloads via the provided download service. Its
  * purpose is to react to global download state changes (e.g. of [BrowserState.downloads])
  * and notify the download service, as needed.
  */
+@Suppress("ComplexMethod")
 class DownloadMiddleware(
     private val applicationContext: Context,
-    private val downloadServiceClass: Class<*>
+    private val downloadServiceClass: Class<*>,
+    coroutineContext: CoroutineContext = Dispatchers.IO,
+    @VisibleForTesting
+    internal val downloadStorage: DownloadStorage = DownloadStorage(applicationContext)
 ) : Middleware<BrowserState, BrowserAction> {
+    private val logger = Logger("DownloadMiddleware")
 
+    private var scope = CoroutineScope(coroutineContext)
+
+    @InternalCoroutinesApi
     override fun invoke(
         store: MiddlewareStore<BrowserState, BrowserAction>,
         next: (BrowserAction) -> Unit,
         action: BrowserAction
     ) {
-        next(action)
         when (action) {
             is DownloadAction.AddDownloadAction -> {
-                val intent = Intent(applicationContext, downloadServiceClass)
-                intent.putExtra(DownloadManager.EXTRA_DOWNLOAD_ID, action.download.id)
-                applicationContext.startService(intent)
+                next(action)
+                scope.launch {
```

Comment:
Here we should check before we launch. No need to launch the coroutine if we don't need to add to the storage.